### ISSUE TITLE
mod_wsgi-standalone random number generator error (CentOS 7)

### ISSUE THREAD

**Issue by genghisken** [GrahamDumpleton/mod_wsgi#628]
HI Graham

I've done a completely new install of mod_wsgi-standalone on a newly recreated CentOS 7 machine.  It installs OK.  However, I'm getting the following error when I try to startup mod_wsgi-express:

```
(forced37) atls::db2 { /files/django_websites/gitrelease/atlasserver }-> ./generate_mod_wsgi_apachectl.sh
Server URL         : http://localhost:8086/
Server Root        : /tmp/atlasforced
Server Conf        : /tmp/atlasforced/httpd.conf
Error Log File     : /tmp/atlasforced/error_log (warn)
Rewrite Rules      : /tmp/atlasforced/rewrite.conf
Environ Variables  : /tmp/atlasforced/envvars
Control Script     : /tmp/atlasforced/apachectl
Request Capacity   : 5 (1 process * 5 threads)
Request Timeout    : 60 (seconds)
Startup Timeout    : 15 (seconds)
Queue Backlog      : 100 (connections)
Queue Timeout      : 45 (seconds)
Server Capacity    : 20 (event/worker), 20 (prefork)
Server Backlog     : 500 (connections)
Locale Setting     : en_GB.UTF-8
[Wed Nov 11 16:53:53.459685 2020] [:crit] [pid 41384] (38)Function not implemented: AH00141: Could not initialize random number generator
```

The generate_mod_wsgi_apachectl.sh script just contains:

```
mkdir -p /tmp/atlasforced
mod_wsgi-express setup-server --working-directory atlasserver --url-alias /sne/atlasforced/static static --application-type module atlasserver.wsgi --server-root /tmp/atlasforced --port 8086 --mount-point /sne/atlasforced

export PYTHONPATH=$(pwd)
/tmp/atlasforced/apachectl start
```

Sounds like I'm missing something crucial in the CentOS 7 operating system install.  Any ideas?  The specific software versions (python 3.7) are:

mod-wsgi-httpd==2.4.41.1
mod-wsgi-standalone==4.7.1

**Comment by GrahamDumpleton** [comment]
Can only suggest doing what is described in:

* https://bz.apache.org/bugzilla/show_bug.cgi?id=63388#c6

So pull down a local copy, edit the ``setup.py`` file and find:

```
    res = os.system('rm -rf build/apr-%(version)s && '
            'tar -x -v -C build -f apr-%(version)s.tar.gz && '
            'cd build/apr-%(version)s && '
            './configure --prefix=%(builddir)s && '
            'make && make install' % dict(builddir=builddir,
            version=APR_VERSION))

    if res:
        raise RuntimeError('Failed to build APR.')
```

Add ``--with-devrandom`` option to the ``configure`` arguments in that command.

Then run:

```
pip install .
```

in the source directory.

**Comment by genghisken** [comment]
Thanks Graham

We've been running CentOS7 systems for years.  I've got another CentOS7 system which is happily running mod_wsgi-standalone with no problems at all.  I'm completely baffled as to why they are different.  Both are running the same version of APR and both the same version of mod_wsgi-standalone.  Both have the same version of Apache installed.  This is a bit of a deployment nightmare for us.  Incidentally I also tried instaling full mod_wsgi and got the same error.  I don't mind attempting the above, but it's an unsustainable solution in the long term for our automated installs.

I think the thing I need to do is go through the two systems with a fine toothcomb and compare differences.  I need to know why it works on one but not the other.

**Comment by GrahamDumpleton** [comment]
Are they definitely using the same mod_wsgi-httpd version by pinning the version? I only updated it to newer Apache version quite recently and it seems more related to newer version of Apache in reports I find about it.

Also, are you creating a wheel and then copying to different boxes or building on each box? It could relate to slightly different kernel versions since how APR libraries configured will be dependent on what the build box uses.

The only other weird explanation I found was this one:

* https://stackoverflow.com/questions/17884176/apache-could-not-initialize-random-number-generator

That is, non ASCII characters in ``PATH``, although why that would cause an issue I have no idea.

**Comment by GrahamDumpleton** [comment]
BTW, if you have Apache installed on the system, why are you using ``mod_wsgi-standalone`` anyway? It should only be used when there is no system Apache and you can't install or use the system Apache for some reason.

**Comment by genghisken** [comment]
On the system where it works OK, I have old apps using the mod_wsgi module tied to python 2.7. (Yeah, yuck.)  In order to get a python 3 app running I had to install mod_wsgi-standalone and proxy from the native apache to mod_wsgi-express.  Works a treat.  I installed mod_wsgi-standalone for the first time on this machine about 5 months ago, then again inside a different python environment just 3 weeks ago.  Both systems have:

mod-wsgi-httpd==2.4.41.1
mod-wsgi-standalone==4.7.1

I'm installing the same architecture on another machine.  (I'm stuck with CentOS7 at the moment.)  Native apache installed and I will proxy to the mod_wsgi-express module as before.  Native httpd will start (on both systems it's apache 2.4.6).

So unless the code was tweaked and the version not updated, or there's some subtle thing in the pip cache on the older machine, I'm really at a loss to explain the difference.  I've googled the problem to death, so also saw that strange ASCII path link!  The older machine also has a slightly older version of Anaconda (inside which my environments live).

**Comment by GrahamDumpleton** [comment]
Hmmm, I got confused, you are on older prior mod_wsgi-httpd version. The version I just released was 2.4.46.1.

As I said before though, if you have system Apache installed, provided you have the ``-devel`` packages for Apache installed, you don't need mod_wsgi-standalone/mod_wsgi-httpd to use mod_wsgi-express behind system Apache, even if the system Apache has different Python version being used with it. You can just use ``pip install mod_wsgi`` and still get a ``mod_wsgi-express`` that should work.

That said, if you are using Anaconda Python then all bets are off. It ships its own versions of some system libraries which can cause conflicts with system ones linked into Apache. Usually it has conflicts when``mod_ssl`` is used, but wouldn't rule out it screwing up random number generation as well.

So two things to pursue.

First is don't, use ``mod_wsgi-standalone``, use just ``mod_wsgi``.

Second is to try and use:

```
pip install https://github.com/GrahamDumpleton/mod_wsgi/archive/develop.zip#egg=mod_wsgi-standalone
```

I just pushed a change to develop branch to ensure it uses the 2.4.46.1 version of mod_wsgi-httpd.

**Comment by genghisken** [comment]
Thanks Graham.  I already tried mod_wsgi.  Got the same error!  But system apache does start OK!

So - this time I explicitly uninstalled mod-wsgi-httpd, which doesn't get uninstalled when you uninstall mod_wsgi-standalone!

THEN - reinstalled mod_wsgi.  Success!

THEN - I also tried your develop branch solution above - also works.  I definitely don't have mod_wsgi installed on the "working" system, so still don't understand the difference, but happy either way.

The reason I went down the mod_wsgi-standalone route was that I wanted to have identical installs on Mac and Linux.  Happy to have something that works though and looks sustainable.  I'll try & get our sysops people to upgrade to a modern OS!

Many, many thanks!!  If it were possible to send you a beer, one would be in the post!
